src/resolver/descriptor.cc

```diff
@@ -0,0 +1,382 @@
+/* ========================================================================== *
+ *
+ * @file resolver/descriptor.cc
+ *
+ * @brief A set of user inputs used to set input preferences and query
+ *        parameters during resolution.
+ *
+ *
+ * -------------------------------------------------------------------------- */
+
+#include <regex>
+
+#include "versions.hh"
+#include "flox/core/exceptions.hh"
+#include "flox/resolver/descriptor.hh"
+
+
+/* -------------------------------------------------------------------------- */
+
+namespace flox::resolver {
+
+/* -------------------------------------------------------------------------- */
+
+/**
+ * @brief Sets either `version` or `semver` on
+ *        a `flox::resolver::ManifestDescriptor`.
+ *
+ * Distinguishes between semver ranges and exact version matchers.
+ * @param desc The descriptor to initialize.
+ * @param version The version description to parse.
+ */
+  static void
+initManifestDescriptorVersion(       ManifestDescriptor & desc
+                             , const std::string        & version
+                             )
+{
+  /* Strip leading/trailing whitespace. */
+  std::string trimmed = trim_copy( version );
+
+  /* Empty is recognized as a glob/_any_ range. */
+  if ( trimmed.empty() )
+    {
+      desc.semver = std::move( trimmed );
+      return;
+    }
+
+  /* Try a quick detection based on first character.
+   * We identify `=` as an explicit _exact version_ match. */
+  switch ( trimmed.at( 0 ) )
+    {
+      case '=':
+        desc.version = std::move( trimmed.substr( 1 ) );
+        break;
+
+      case '*':
+      case '~':
+      case '^':
+      case '>':
+      case '<':
+        desc.semver = std::move( trimmed );
+        break;
+
+      default:
+        /* If it's a valid semver or a date then it's not a range. */
+        if ( versions::isSemver( trimmed ) || versions::isDate( trimmed ) ||
+             ( ! versions::isSemverRange( trimmed ) )
+           )
+          {
+            desc.version = std::move( trimmed );
+          }
+        else /* Otherwise, assume a range. */
+          {
+            desc.semver = std::move( trimmed );
+          }
+        break;
+    }
+}
+
+
+/* -------------------------------------------------------------------------- */
+
+/** @brief Get a `flox::resolver::AttrPathGlob` from a string if necessary. */
+  static AttrPathGlob
+maybeSplitAttrPathGlob( const ManifestDescriptorRaw::AbsPath & absPath )
+{
+  if ( std::holds_alternative<AttrPathGlob>( absPath ) )
+    {
+      return std::get<AttrPathGlob>( absPath );
+    }
+  AttrPathGlob   glob;
+  flox::AttrPath path = splitAttrPath( std::get<std::string>( absPath ) );
+  size_t         idx  = 0;
+  for ( const auto & part : path )
+    {
+      /* Treat `null' or `*' in the second element as a glob. */
+      if ( ( idx == 1 ) && ( ( part == "null" ) || ( part == "*" ) ) )
```

Comment:
```suggestion
      /* Treat `null' or `*' as a glob. */
      /* TODO we verify that only the second option is a glob elsewhere, but we could do that here instead */
      if ( ( ( part == "null" ) || ( part == "*" ) ) )
```

Reply:
Since we've already done it in the block above I'm not sure that it makes sense to repeat those checks here

Reply:
What block are these checks done in?

Reply:
https://github.com/flox/pkgdb/pull/76/files#diff-c7d6d0eb80d5d13a6a8424de32efea1281bfd24fefc950e83dedb274d6cec038R144

Reply:
That line only checks for nullopt, not "null" or "*", which should be done here. If you tweak the manifest test like this:
```
diff --git a/tests/manifest.cc b/tests/manifest.cc
index 706c314..86a61f4 100644
--- a/tests/manifest.cc
+++ b/tests/manifest.cc
@@ -252,7 +252,7 @@ test_parseManifestDescriptor_path0()
 {

   flox::resolver::ManifestDescriptorRaw raw = R"( {
-    "absPath": "legacyPackages.null.hello"
+    "absPath": "legacyPackages.null.null"
   } )"_json;

   flox::resolver::ManifestDescriptor descriptor( raw );
@@ -262,7 +262,7 @@ test_parseManifestDescriptor_path0()
   EXPECT( ! descriptor.systems.has_value() );
   EXPECT( ! descriptor.stability.has_value() );
   EXPECT( descriptor.path.has_value() );
-  EXPECT( ( * descriptor.path ) == ( flox::AttrPath { "hello" } ) );
+  EXPECT( ( * descriptor.path ) == ( flox::AttrPath { "null" } ) );

   return true;
 }
```
It will currently pass. With my patch, it will fail with:
```
Testing: tests/manifest
  ERROR: parseManifestDescriptor_path0: `absPath' may only have a glob as its second element
```

Reply:
I see what you mean. I was confused about "checking" to mean "exception handling".
I think you're talking about the parsing?

That's a good catch. `["legacyPackages",null,"null"]` for a package named `null`.

Reply:
By the TODO I mean it might make more sense to verify that only the second element is a glob here, rather than higher up in multiple places. But whatever you prefer 🤷 

Regardless, "legacyPackages.null.null" and "legacyPackages.null.*" need to ERROR, and currently they aren't.

Reply:
There is technically a package named `legacyPackages.null.null` by our spec which is what I thought you were highlighting.

There's a package named `null`. Your test case shows that I didn't properly handle that edge case, so that's definitely something to fix in a follow up PR.

Thanks for highlighting the gap :v: